lua/ulx/modules/urs_server.lua

```diff
@@ -1,193 +1,276 @@
+
+    return result
+end
+local cacheCheck = URS.CacheCheck
+
+function URS.Check(ply, restrictionType, what)
+    local cachedResult = cachedCheck( ply, restrictionType, what )
+    if cachedResult ~= nil then
+        PrintRestricted( ply, restrictionType, what )
+        return cachedResult
+    end
+
+    what = stringLower(what)
+    local restrictionTypePlural = restrictionType .. "s"
+    local group = ply:GetUserGroup()
+
+    local restriction = false
+
+    local ursRestrictions = rawget( URS, "restrictions" )
+    local typeRestrictions = ursRestrictions and rawget( ursRestrictions, restrictionType )
+    local theseRestrictions = typeRestrictions and rawget( typeRestrictions, what )
+
+    if theseRestrictions then
+        restriction = theseRestrictions
+    end
+
+    if restriction then
+        local hasGroup = rawget( restriction, group )
+        local hasPlayer = rawget( restriction, ply:SteamID() )
+
+        if rawget( restriction, "*" ) then
+            if not ( hasGroup or hasPlayer ) then
+                PrintRestricted( ply, restrictionType, what )
+                return cacheCheck( ply, restrictionType, what, false )
+            end
+        elseif hasGroup or hasPlayer then
+            PrintRestricted( ply, restrictionType, what )
+            return cacheCheck( ply, restrictionType, what, false )
+        end
+    end
+
+    local allRestrictions = rawget( ursRestrictions, "all" )
+    local allTypeRestrictions = allRestrictions and rawget( allRestrictions, restrictionType )
+    local hasGroup = allTypeRestrictions and rawget( allTypeRestrictions, group )
+
+    if hasGroup then
+        tsayError(ply, "Your rank is restricted from all " .. restrictionTypePlural)
+        return cacheCheck( ply, restrictionType, what, false )
+    end
+
+    local ursTypes = rawget( URS, "types" )
+    local ursTypeLimits = rawget( ursTypes, "limitsMap" )
+    local limitsHasType = rawget( ursTypeLimits, restrictionType )
+    if not limitsHasType then return end
+
+    local ursLimits = rawget( URS, "limits" )
+    local typeLimits = rawget( ursLimits, restrictionType )
+    if not typeLimits then return end
+
+    local playerTypeLimit = rawget( typeLimits, ply:SteamID() )
+    local groupTypeLimit = rawget( typeLimits, group )
+    if not playerTypeLimit and not groupTypeLimit then return end
+
+    -- TODO: Should this be an elseif? Shouldn't both cases be possible?
+    if playerTypeLimit then
+        if ply:GetCount(restrictionTypePlural) >= playerTypeLimit then
+            ply:LimitHit( restrictionTypePlural )
+            return false
+        end
+    elseif groupTypeLimit then
+        if ply:GetCount(restrictionTypePlural) >= groupTypeLimit then
+            ply:LimitHit( restrictionTypePlural )
+            return false
+        end
+    end
+    if overwriteSbox then
+        return true -- Overwrite sbox limit (ours is greater)
+    end
+end
+local Check = URS.Check
+
+timer.Simple( 0.1, function()
```

Comment:
I feel like this can be optimized a lot, maybe not too needed but it's bad in its current state